### PR TITLE
Add preventing self delegation logic

### LINKED ISSUE #363: Prevent self delegation
Delegation to myself must be rejected.

1. Add check logic in `apply_delegate()`
2. Add a test that checks a self delegation is successfully rejected.

### PR DESCRIPTION
**Pull request by just-do-halee**
Fixes #363

### CODE REVIEW

File: common/src/verify.rs [hunk 1]
```diff
@@ -372,63 +372,46 @@ impl CommitSequenceVerifier {
...
                             return Err(Error::InvalidArgument(
-                                format!("invalid extra-agenda transaction timestamp: expected larger than or equal to the last transaction timestamp {}, got {}", last_extra_agenda_timestamp, tx.data.timestamp)
+                                format!("invalid extra-agenda transaction timestamp: expected larger than or equal to the last transaction timestamp {last_extra_agenda_timestamp}, got {timestamp}")
                             ));
                         }
-                        *last_extra_agenda_timestamp = tx.data.timestamp;
+                        *last_extra_agenda_timestamp = timestamp;
                     }
-                    ExtraAgendaTransaction::Report(_tx) => unimplemented!(),
+                    _ => unreachable!(),
```
Comment (junha1) [inline common/src/verify.rs:420]: It is reachable. Note that you are missing `PhaseMismatch`.
I don't get the point of your refactoring.

File: common/src/reserved.rs [hunk 2]
```diff
@@ -507,6 +513,43 @@ mod tests {
...
+        let tx = TxDelegate { data, proof };
+
+        let new_state = state.apply_delegate(&tx);
+
+        // then
+        //
+        if let Err(s) = new_state {
+            assert_eq!(
+                s,
+                format!("delegator and delegatee are the same: {}", delegator.name)
```
Comment (junha1) [inline common/src/reserved.rs:546]: We do not assert on the error message itself (it's too fragile)
Unfortunately, the only thing we can do is just checking whether it is `Err`.

File: common/src/reserved.rs [hunk 3]
```diff
@@ -507,6 +513,43 @@ mod tests {
...
+            );
+        } else {
+            panic!("Unexpected error");
```
Comment (junha1) [inline common/src/reserved.rs:549]: Try more helpful error message.
How about `it must fail when the delegator and the delegatee are the same`

File: common/src/reserved.rs [hunk 4]
```diff
@@ -507,6 +513,43 @@ mod tests {
...
+            chain_name: state.genesis_info.chain_name.clone(),
+        };
+        let proof = TypedSignature::sign(&data, &delegator_private_key).unwrap();
+
+        let tx = TxDelegate { data, proof };
+
+        let new_state = state.apply_delegate(&tx);
+
+        // then
+        //
```
Comment (TomTaehoonKim) [inline common/src/reserved.rs:542]: How about deleting this line?

File: common/src/reserved.rs [hunk 5]
```diff
@@ -507,6 +513,43 @@ mod tests {
         );
     }
 
+    #[test]
+    fn test_apply_delegate_on_consensus_failure() {
+        // given
+        setup_test();
+        let (mut state, keys) = generate_standard_genesis(3);
+
+        // delegator: member-0000, delegatee: member-0002
```
Comment (TomTaehoonKim) [inline common/src/reserved.rs:522]: How about deleting this line? I think copy and pasting this line is not necessary and misleading.

**Pull request merged**